pwiz_tools/Skyline/Controls/Graphs/DetectionPlotData.cs

```diff
@@ -0,0 +1,417 @@
+            private StackWorker<DataRequest> _stackWorker;
+            private CancellationTokenSource _tokenSource;
+            private Action<DetectionPlotData> _callback;
+            private object _statusLock = new object();
+
+            // This is max number of replicates that the cache will store before starting to purge 
+            // the old datasets. Number of replicates is a good proxy for the amount of memory used 
+            // by a dataset.
+            private const int CACHE_CAPACITY = 200;
+
+            public event Action<CacheStatus> StatusChange;
+            public event Action<int> ReportProgress;
+
+            public enum CacheStatus { idle, processing, error, canceled }
+
+            private CacheStatus _status;
+            public CacheStatus Status
+            {
+                get => _status;
+                set
+                {
+                    _status = value;
+                    StatusChange?.Invoke(_status);
+                }
+            }
+
+            public DetectionDataCache()
+            {
+                _stackWorker = new StackWorker<DataRequest>(null, CacheData);
+                //single worker thread because we do not need to paralellize the calculations, 
+                //only to offload them from the UI thread
+                _stackWorker.RunAsync(1, @"DetectionsDataCache");
+                _tokenSource = new CancellationTokenSource();
+                _datas = new ConcurrentQueue<DetectionPlotData>();
+                Status = CacheStatus.idle;
+            }
+
+            public bool TryGet(SrmDocument doc, float qValue, Action<DetectionPlotData> callback,  out DetectionPlotData data)
+            {
+                //Debug.WriteLine($@"Received request for q {qValue}");
+                data = DetectionPlotData.INVALID;
+                var request = new DataRequest() { qValue = qValue};
+                if (ReferenceEquals(doc, _document))
+                {
+                    data = Get(request) ?? DetectionPlotData.INVALID;
+                    if (data.IsValid)
+                    {
+                        //Debug.WriteLine($@"Found cached data for q {qValue}");
+                        return true;
+                    }
+                    //Debug.WriteLine($@"No cached data for q {qValue}, adding to the queue");
+                    _callback = callback;
+                    _stackWorker.Add(request);
+                }
+                else
+                {
+                    _document = doc;
+                    //Debug.WriteLine($@"New document request for q {qValue}, starting cancel.");
+                    new Task(CancelWorker, new object[] { request, false }).Start(); 
+                }
+
+                return false;
+            }
+
+            public void Cancel()
+            {
+                new Task(CancelWorker, new object[]{null, true}).Start();
+
+            }
+
+            private void CancelWorker(object param)

```

Comment:
It would be easier to understand this function if it took two arguments: DataRequest, bool

Reply:
Agree. I didn't realize I can use lambda in the Task constructor.